spring-cloud-huawei-nacos/src/main/java/com/huaweicloud/nacos/authentication/NacosInstanceService.java

```diff
@@ -0,0 +1,136 @@
+import com.alibaba.nacos.api.exception.NacosException;
+import com.google.common.cache.Cache;
+import com.google.common.cache.CacheBuilder;
+import com.google.common.util.concurrent.UncheckedExecutionException;
+import com.huaweicloud.common.disovery.InstanceIDAdapter;
+import com.huaweicloud.governance.authentication.Const;
+import com.huaweicloud.governance.authentication.MicroserviceInstanceService;
+import com.huaweicloud.nacos.common.NacosMicroservice;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.cloud.client.ServiceInstance;
+
+import java.beans.PropertyDescriptor;
+import java.util.Map;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+public class NacosInstanceService implements MicroserviceInstanceService {
+    private static final Logger LOGGER = LoggerFactory.getLogger(NacosInstanceService.class);
+
+    private final NacosDiscoveryProperties properties;
+
+    private final NacosServiceDiscovery serviceDiscovery;
+
+    private final NacosRegistration registration;
+
+    private String publickey = "";
+
+    public NacosInstanceService(NacosDiscoveryProperties properties, NacosRegistration registration,
+        NacosServiceDiscovery serviceDiscovery) {
+        this.properties = properties;
+        this.registration = registration;
+        this.serviceDiscovery = serviceDiscovery;
+    }
+
+    @Override
+    public void setPublickey(String publicKeyEncoded) {
+        publickey = publicKeyEncoded;
+        properties.getMetadata().put(Const.INSTANCE_PUBKEY_PRO, publicKeyEncoded);
+    }
+
+    @Override
+    public String getInstanceId() {
+        ServiceInstance serviceInstance = null;
+        try {
+            serviceInstance = serviceDiscovery.getInstances(registration.getServiceId())
+                    .stream()
+                    .filter(instance -> instance.getMetadata().containsValue(publickey))
+                    .findAny()
+                    .get();
+        } catch (NacosException e) {
+            throw new RuntimeException(e);
+        }
+        return InstanceIDAdapter.instanceId(serviceInstance);
+    }
+
+    @Override
+    public String getServiceId() {
+        return registration.getServiceId();
+    }
+
+    @Override
+    public String getPublicKeyFromInstance(String instanceId, String serviceId) {
+        ServiceInstance instances = getMicroservice(serviceId, instanceId);
+        if (instances != null) {
+            return instances.getMetadata().get(Const.INSTANCE_PUBKEY_PRO);
+        } else {
+            LOGGER.error("not instance found {}-{}, maybe attack", instanceId, serviceId);
+            return "";
+        }
+    }
+
+    @Override
+    public Object getFieldValue(String propertyName, String serviceId, String instanceId) throws Exception {
+        ServiceInstance serviceInstance = getMicroservice(serviceId, instanceId);
+        NacosMicroservice nacosMicroservice = new NacosMicroservice();
```

Comment:
Seams not correct.  Should get from properties. Maybe you can abstract all to getProperty and nacos only get extened properties.

Reply:
abstract all to getProperty method.